fix: Reverse row selection should respect the row group index

zhuqi-lucas opened the pull request:
## Which issue does this PR close?

- Closes [#19535](https://github.com/apache/datafusion/issues/19535)
## Rationale for this change

Reverse row selection should respect the row group index, this PR will fix the issue.

## What changes are included in this PR?

Reverse row selection should respect the row group index, this PR will fix the issue.


## Are these changes tested?

Yes

## Are there any user-facing changes?

No

adriangb commented on datafusion/datasource-parquet/src/sort.rs:
```diff
@@ -31,22 +31,34 @@ use std::collections::HashMap;
     let mut rg_row_ranges: Vec<(usize, usize, usize)> =
-        Vec::with_capacity(rg_metadata.len());
+        Vec::with_capacity(row_groups_to_scan.len());
     let mut current_row = 0;
```
Is the tracking for `current_row` still correct? We used to scan every row so I think it was correct, but now we if we skip a row group won't `current_row` be incorrect globally?

adriangb commented on datafusion/datasource-parquet/src/sort.rs:
```diff
@@ -153,7 +166,10 @@ mod tests {
             RowSelection::from(vec![RowSelector::select(50), RowSelector::skip(250)]);
 
-        let reversed = reverse_row_selection(&selection, &metadata).unwrap();
+        let row_groups_to_scan = vec![0, 1, 2];
```
Isn't `row_groups_to_scan` derived from the `RowSelection`? So in this case it should be `vec![0]`?

zhuqi-lucas replied:
Ok, i got it now @adriangb , i will try to address it, thanks!

zhuqi-lucas replied:
Addressed in latest PR.

adriangb commented on datafusion/datasource-parquet/src/opener.rs:
Maybe we should be writing tests as:

```rust
/// Test helper
fn reverse_access_plan(plan: ParquetAccessPlan, metadata: &ParquetMetaData) -> RowSelection {...}

/// Test code
let plan = ParquetAccessPlan...
let expected = RowSelection ...
let reversed = reverse_access_plan(plan, metadata)
assert_eq!(expected, reversed)
```

zhuqi-lucas replied:
Great point @adriangb , this looks better.

zhuqi-lucas commented:
Thank you @adriangb for review, i have addressed the comments and used 100% real cases to test now which building a PreparedAccessPlan and calling PreparedAccessPlan::reverse for all tests in sort.rs.

zhuqi-lucas commented on datafusion/datasource-parquet/src/sort.rs:
```diff
@@ -101,67 +120,165 @@ pub fn reverse_row_selection(
+
+        // If no selection originally, after reversal should still select all rows
+        if let Some(selection) = reversed_plan.row_selection {
+            let total_selected: usize = selection
```
Updated:

```rust
 #[test]
    fn test_prepared_access_plan_reverse_simple() {
        // Test: all row groups are scanned, no row selection
        let metadata = create_test_metadata(vec![100, 100, 100]);

        let access_plan = ParquetAccessPlan::new_all(3);
        let rg_metadata = metadata.row_groups();

        let prepared_plan =
            PreparedAccessPlan::from_access_plan(access_plan, rg_metadata)
                .expect("Failed to create PreparedAccessPlan");

        // Verify original plan
        assert_eq!(prepared_plan.row_group_indexes, vec![0, 1, 2]);

        // No row selection originally due to scanning all rows
        assert_eq!(prepared_plan.row_selection, None);

        let reversed_plan = prepared_plan
            .reverse(&metadata)
            .expect("Failed to reverse PreparedAccessPlan");

        // Verify row groups are reversed
        assert_eq!(reversed_plan.row_group_indexes, vec![2, 1, 0]);

        // If no selection originally, after reversal should still select all rows,
        // and the selection should be None
        assert_eq!(reversed_plan.row_selection, None);
    }
```
